internal/database/offline/sqlutil/export.go

```diff
@@ -13,24 +13,50 @@ import (
 	"github.com/oom-ai/oomstore/pkg/oomstore/types"
 )
 
+type QueryExportResults func(ctx context.Context, dbOpt dbutil.DBOpt, opt offline.ExportOpt, query string, args []interface{}) (<-chan types.ExportRecord, <-chan error)
+
+type DoExportOpt struct {
+	offline.ExportOpt
+	QueryResults QueryExportResults
+}
+
 func Export(ctx context.Context, db *sqlx.DB, opt offline.ExportOpt, backend types.BackendType) (<-chan types.ExportRecord, <-chan error) {
+	dbOpt := dbutil.DBOpt{
+		Backend: backend,
+		SqlxDB:  db,
+	}
+	doJoinOpt := DoExportOpt{
+		ExportOpt:    opt,
+		QueryResults: sqlxQueryExportResults,
+	}
+	return DoExport(ctx, dbOpt, doJoinOpt)
+}
+
+func DoExport(ctx context.Context, dbOpt dbutil.DBOpt, opt DoExportOpt) (<-chan types.ExportRecord, <-chan error) {
 	var (
-		stream = make(chan types.ExportRecord)
-		errs   = make(chan error, 1) // at most 1 error
+		emptyStream = make(chan types.ExportRecord)
+		errs        = make(chan error, 1) // at most 1 error
 	)
+	defer close(emptyStream)
+	defer close(errs)
 
-	query, args, err := buildExportQuery(opt, backend)
+	query, args, err := buildExportQuery(dbOpt, opt.ExportOpt)
 	if err != nil {
-		defer close(stream)
-		defer close(errs)
 		errs <- errdefs.WithStack(err)
-		return stream, errs
+		return emptyStream, errs
 	}
 
+	return opt.QueryResults(ctx, dbOpt, opt.ExportOpt, query, args)
+}
+
+func sqlxQueryExportResults(ctx context.Context, dbOpt dbutil.DBOpt, opt offline.ExportOpt, query string, args []interface{}) (<-chan types.ExportRecord, <-chan error) {
```

Comment:
`sqlxQueryExportResults` supports bigquery, would it be more appropriate to remove the sqlx prefix?

Reply:
no it doesn't support bigquery, bigquery has its own query function

Reply:
my bad.